### PR TITLE
fix: validate google submit receipt

### PR DESCRIPTION
**Pull request by clD11**
### Summary

This PR fixes the validate google submit receipt checks and also removes the unnecessary dependency on the awa lib. To fully remove the lib we will also need to refactor validate IOS in a separate PR.

**Linking Conditions:**

We should not allow a user to link when their subscription is either in a pending or expired state. 

### Type of Change

- [ ] Product feature
- [x] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security and/or privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: services/skus/controllers.go [hunk 1]
```diff
@@ -1590,14 +1590,10 @@ func handleReceiptErr(err error) *handlers.AppError {
 	}
 
 	switch {
-	case errors.Is(err, errPurchaseFailed):
-		result.ErrorCode = purchaseFailedErrCode
+	case errors.Is(err, errPurchaseExpired):
+		result.ErrorCode = purchaseExpiredErrCode
```
Comment (clD11) [inline services/skus/controllers.go:1594]: note: I checked the skus SDK and searched the Brave github org and could not see anything depending on this error code so it looks safe to change/add.

File: services/skus/receipt.go [hunk 2]
```diff
@@ -14,53 +14,24 @@ import (
...
-	androidCancelReasonDeveloper int64 = 3
-
-	purchasePendingErrCode       = "purchase_pending"
-	purchaseDeferredErrCode      = "purchase_deferred"
-	purchaseStatusUnknownErrCode = "purchase_status_unknown"
-	purchaseFailedErrCode        = "purchase_failed"
-	purchaseValidationErrCode    = "validation_failed"
+	purchasePendingErrCode    = "purchase_pending"
+	purchaseExpiredErrCode    = "purchase_expired"
+	purchaseValidationErrCode = "validation_failed"
```
Comment (clD11) [inline services/skus/receipt.go:28]: note: this is shared by both ios and android, we may want to update this as its not very descriptive and maybe confusing.

Comment (pavelbrm) [reply services/skus/receipt.go:28]: I'd opt for doing it as a follow-up.

File: services/skus/receipt.go [hunk 3]
```diff
@@ -14,53 +14,24 @@ import (
 	"time"
 
 	"github.com/awa/go-iap/appstore"
-	"github.com/awa/go-iap/playstore"
-	"google.golang.org/api/androidpublisher/v3"
-
 	appctx "github.com/brave-intl/bat-go/libs/context"
 	"github.com/brave-intl/bat-go/libs/logging"
+	"google.golang.org/api/androidpublisher/v3"
```
Comment (pavelbrm) [inline services/skus/receipt.go:19]: nit/suggestion: Would something like this make import paths slightly shorter and manageable?

```suggestion
	androidpub "google.golang.org/api/androidpublisher/v3"
```

File: services/skus/receipt.go [hunk 4]
```diff
@@ -83,17 +54,36 @@ func (dt *dumpTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	return resp, rtErr
 }
 
+type androidPublisher struct {
```
Comment (pavelbrm) [inline services/skus/receipt.go:57]: suggestion: It does look like a `playStoreClient` to me. `androidPublisher` does not fully describe what it does (we certainly don't publish anything).

File: services/skus/receipt.go [hunk 5]
```diff
@@ -145,54 +137,45 @@ func (v *receiptVerifier) validateApple(ctx context.Context, req model.ReceiptRe
...
+	return req.Blob, nil
+}
 
-	case androidPaymentStatePendingDeferred:
-		return "", errPurchaseDeferred
-	default:
-		return "", errPurchaseStatusUnknown
+func isExpired(expTimeMills int64, now time.Time) bool {
+	return now.UnixMilli() > expTimeMills
+}
```
Comment (pavelbrm) [inline services/skus/receipt.go:165]: suggestion: The name seems a bit too generic. We could make it more specific in the context of time it is dealing with, or the task it performs:

1. A more time-specific option:
```golang
func hasExpiredMilli(now, target int64) bool {
	return now > target
}
```

2. A more task-specific name:
```golang
func hasPlaySubExpired(now time.Time, targetMilli int64) bool {
	return now.Milli() > targetMilli
}
```

File: services/skus/receipt.go [hunk 6]
```diff
@@ -145,54 +137,45 @@ func (v *receiptVerifier) validateApple(ctx context.Context, req model.ReceiptRe
...
+	androidPaymentStatePendingDeferred int64 = 3
+)
+
+func isSubPurchasePending(resp *androidpublisher.SubscriptionPurchase) bool {
+	// The payment state is not present for canceled or expired subscriptions.
+	if resp.PaymentState != nil {
+		return *resp.PaymentState == androidPaymentStatePending || *resp.PaymentState == androidPaymentStatePendingDeferred
 	}
+
+	return false
```
Comment (pavelbrm) [inline services/skus/receipt.go:178]: Consider:

```suggestion
func isSubPurchasePending(resp *androidpublisher.SubscriptionPurchase) bool {
	// The payment state is not present for canceled or expired subscriptions.
	if resp.PaymentState == nil {
		return false
	}

	s := *resp.PaymentState

	return s == androidPaymentStatePending || s == androidPaymentStatePendingDeferred
}
```

1. Less indentation.
2. One dereference for the worst case.

**Review by pavelbrm — COMMENTED** [review]
Looks much cleaner! ✨ 💅 

Left some minor feedback. Let me know when it's out of draft.

File: services/skus/receipt.go (hunk 5 above)
Comment (clD11) [reply services/skus/receipt.go]: update so its more specific

File: services/skus/receipt.go (hunk 4 above)
Comment (clD11) [reply services/skus/receipt.go:57]: This is simply just a wrapper around Google `androidpublisher` api which is their naming convention, the intention is to simply wrap it and not imply there is any extra functionality etc. IMO when looking up docs it is also a little clear if it simply wrap their naming.

File: services/skus/receipt.go (hunk 6 above)
Comment (clD11) [reply services/skus/receipt.go]: I originally had your suggestion but changed it to its current form for the reasons we discussed. Updating it to this structure.

File: services/skus/receipt_test.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/brave-intl/bat-go/libs/ptr"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/api/androidpublisher/v3"
+)
+
+func TestIsExpired(t *testing.T) {
```
Comment (pavelbrm) [inline services/skus/receipt_test.go:12]:
```suggestion
func TestIsSubPurchaseExpired(t *testing.T) {
```

File: services/skus/receipt_test.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		tc := tests[i]
+
+		t.Run(tc.name, func(t *testing.T) {
+			actual := isSubPurchaseExpired(tc.given.expTimeMills, tc.given.now)
+			assert.Equal(t, tc.exp, actual)
+		})
+	}
+}
+
+func TestIsPending(t *testing.T) {
```
Comment (pavelbrm) [inline services/skus/receipt_test.go:58]:
```suggestion
func TestIsSubPurchasePending(t *testing.T) {
```

**Review by pavelbrm — APPROVED** [review]
Loving it, much better. 🤩

**Pull request merged**